x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/http/retry/RetryingHttpSender.java

```diff
@@ -0,0 +1,158 @@
+        Logger logger,
+        RetrySettings retrySettings,
+        ThreadPool threadPool,
+        Executor executor
+    ) {
+        this.sender = Objects.requireNonNull(sender);
+        this.throttlerManager = Objects.requireNonNull(throttlerManager);
+        this.logger = Objects.requireNonNull(logger);
+        this.retrySettings = Objects.requireNonNull(retrySettings);
+        this.threadPool = Objects.requireNonNull(threadPool);
+        this.executor = Objects.requireNonNull(executor);
+    }
+
+    private class InternalRetrier extends RetryableAction<InferenceResults> {
+        private final HttpRequestBase request;
+        private final ResponseHandler responseHandler;
+
+        InternalRetrier(HttpRequestBase request, ResponseHandler responseHandler, ActionListener<InferenceResults> listener) {
+            super(
+                logger,
+                threadPool,
+                retrySettings.getSettings().initialDelay(),
+                retrySettings.getSettings().maxDelayBound(),
+                retrySettings.getSettings().timeoutValue(),
+                listener,
+                executor
+            );
+            this.request = request;
+            this.responseHandler = responseHandler;
+        }
+
+        @Override
+        public void tryAction(ActionListener<InferenceResults> listener) {
+            ActionListener<HttpResult> responseListener = ActionListener.wrap(result -> {
+                try {
+                    responseHandler.validateResponse(throttlerManager, logger, request, result);
+                    InferenceResults inferenceResults = responseHandler.parseResult(result);
+
+                    listener.onResponse(inferenceResults);
+                } catch (Exception e) {
+                    logException(request, result, responseHandler.getRequestType(), e);
+                    listener.onFailure(e);
+                }
+            }, e -> {
+                logException(request, responseHandler.getRequestType(), e);
+                listener.onFailure(transformIfRetryable(e));
+            });
+
+            sender.send(request, responseListener);
+        }
+
+        @Override
+        public boolean shouldRetry(Exception e) {
+            if (e instanceof RetryException) {
+                return ((RetryException) e).shouldRetry();
+            }
+
+            return false;
+        }
+
+        /**
+         * If the connection gets closed by the server or because of the connections time to live is exceeded we'll likely get a
+         * {@link org.apache.http.ConnectionClosedException} exception which is a child of IOException. For now,
+         * we'll consider all IOExceptions retryable because something failed while we were trying to send the request
+         * @param e the Exception received while sending the request
+         * @return a {@link RetryException} if this exception can be retried
+         */
+        private Exception transformIfRetryable(Exception e) {
```

Comment:
TODO needs to handle task rejection and timeout errors when the queuing logic fails.